### PR TITLE
chore(web): replace lerna with npm workspaces and ts projects 🎡

### LINKED ISSUE #6320: chore(common): move from lerna to alternative monorepo solution
EDIT: this has morphed a bit, hijacking to turn this into lerna-replacement issue. See discussion in comments.

---

We need to update the lockfile version on a dev machine, and deal with the deprecated packages.

```
[18:14:21][Step 1/2] npm WARN old lockfile 
[18:14:21][Step 1/2] npm WARN old lockfile The package-lock.json file was created with an old version of npm,
[18:14:21][Step 1/2] npm WARN old lockfile so supplemental metadata must be fetched from the registry.
[18:14:21][Step 1/2] npm WARN old lockfile 
[18:14:21][Step 1/2] npm WARN old lockfile This is a one-time fix-up, please be patient...
[18:14:21][Step 1/2] npm WARN old lockfile 
[18:14:31][Step 1/2] npm WARN EBADENGINE Unsupported engine {
[18:14:31][Step 1/2] npm WARN EBADENGINE   package: '@keymanapp/auto-history-action@1.1.3',
[18:14:31][Step 1/2] npm WARN EBADENGINE   required: { node: '12' },
[18:14:31][Step 1/2] npm WARN EBADENGINE   current: { node: 'v16.13.2', npm: '8.1.2' }
[18:14:31][Step 1/2] npm WARN EBADENGINE }
[18:14:33][Step 1/2] npm WARN deprecated @zeit/ncc@0.21.0: @zeit/ncc is no longer maintained. Please use @vercel/ncc instead.
```

and

```
[18:17:40][Step 1/2] warning: LF will be replaced by CRLF in HISTORY.md.
[18:17:40][Step 1/2] The file will have its original line endings in your working directory
[18:17:42][Step 1/2] npm WARN old lockfile 
[18:17:42][Step 1/2] npm WARN old lockfile The package-lock.json file was created with an old version of npm,
[18:17:42][Step 1/2] npm WARN old lockfile so supplemental metadata must be fetched from the registry.
[18:17:42][Step 1/2] npm WARN old lockfile 
[18:17:42][Step 1/2] npm WARN old lockfile This is a one-time fix-up, please be patient...
[18:17:42][Step 1/2] npm WARN old lockfile 
[18:19:01][Step 1/2] npm WARN deprecated read-package-tree@5.3.1: The functionality that this package provided is now in @npmcli/arborist
[18:19:01][Step 1/2] npm WARN deprecated har-validator@5.1.5: this library is no longer supported
[18:19:02][Step 1/2] npm WARN deprecated uuid@3.4.0: Please upgrade  to version 7 or higher.  Older versions may use Math.random() in certain circumstances, which is known to be problematic.  See https://v8.dev/blog/math-random for details.
[18:19:02][Step 1/2] npm WARN deprecated request@2.88.2: request has been deprecated, see https://github.com/request/request/issues/3142
```

https://build.palaso.org/viewLog.html?buildId=309145&buildTypeId=Keyman_TriggerReleaseBuildsMaster&tab=buildLog&_focus=480

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #6320.

Part 1 of moving from lerna to a simpler, maintained monorepo solution,  using TypeScript Projects and NPM Workspaces.

There is more work to be done here. At this point, KeymanWeb builds unminified and runs without errors (locally at least!), but the built file is substantially different, mostly in include order.

Using TypeScript Projects, we move away from the need to run build scripts in various locations for almost all the Typescript modules. `npm install` or `npm ci` is needed just once at the top level of the repo. The TypeScript project references mirror the npm references. A discussion may be worthwhile on how we link the two (or if we do) -- this seems to be an open question for TypeScript currently.

On my machine, this reorg reduces the KeymanWeb build from several minutes to about 15 seconds! There may be more to do though.

## TODO

- [x] Embedded versions and tests have not been verified. #6597 #6602 #6606 #6608
- [x] developer/server is not yet verified. #6612
- [x] developer/js (needs a rename!) is not yet verified. #6610 #6613
- [x] Currently, the predictive-text folder needs refactoring to move the construction of the worker wrapper out of the Predictive Text build and into the final assembly of keymanweb.js (as it should be valid to run it as a separate .js anyway). See #6533.
- [x] several ts projects use inconsistent output folders. #6586
- [x] eliminate version numbers from package.json - see #6539
- [x] Most of the `<reference>` paths need to be re-verified. Ideally there should be no references outside the current module for any given .ts. #6616
- [x] Use `npm ci` instead of `npm install` (ref #6196)
- [x] resources/web-environment should be in common/web. (and how does it overlap with keyman-version?) - see #6617
- [x] it may be possible to generate a .d.ts for models/types so that we can use a consistent reference for those as well
- [x] fix up the publish code for npm modules - see #6618
- [x] fixup unit tests for predictive text - see #6622
- [x] fixup source maps

For the build script and other cleanup phases (not part of this PR chain):
- [x] merge web-environment and keyman-version
- [x] cleanup "poor man's module mode" for tests from #6597.

Note, remaining TODO items moved to #6268 (may be moved again from there).

Whew, that's most of the stuff I noticed!

### CODE REVIEW

File: web/source/build.sh
Comment (jahorton) [inline web/source/build.sh:57]: The CI for Web fell over here, likely due to this change.

**Comment by jahorton** [comment]
When I cloned a new copy of the repo to examine this - so that it doesn't pollute my main working area - I got this error during the Web build:

```
10:06:42 AM - Failed to parse file '../resources/web-environment/tsconfig.json': C:/keymanapp/keyman-reorg/resources/web-environment/environment.inc.ts does not exist.
```

Obviously, not the hardest thing to resolve.  But something of note.

Next issue:

```
/c/keymanapp/keyman-reorg/web/source
cat: ../../common/predictive-text/build/intermediate/index.js: No such file or directory
```

It sounds as if the WebWorker-internal workspace is not being built.  (It has its own ts-config.json.)  *Kinda* need that for predictive text to work properly... and this means that Web isn't being fully compiled as of yet.

My thoughts toward a possible resolution:
- We may need something similar to your `es6-shim` wrapper for the WebWorker internals.
- To facilitate *that*...
    - `common/predictive-text/worker` should become its own, standalone package - perhaps `lm-worker-internals`?
    - New package:  `wrapped-worker`:
        - A local-only NPM package that 'builds' by running the `build-wrapped-worker` function from `common/predictive-text/build.sh`, distributing its results
        - Obviously, depends on `lm-worker-internals`.
        - Plays a similar role to your `es6-shim` wrapper. 
    - From there, `lexical-model-layer` should be able to `path`-link to the `wrapped-worker` TS project.

It's quite possible that this could be simplified; I just based this on my current level of understanding of the new "toys" you're "playing" with.  I'm fairly certain this breakdown would work if what you're already doing works for es6-shim.

------

Both issues above became evident rapidly because it's a clean workspace (fresh checkout).

**Comment by mcdurdin** [comment]
Thanks for the quick review @jahorton. Some quick notes:

* The way I understand it, `npm ci` is broadly the same as `npm install` in terms of bandwidth requirements. The difference is that `npm install` can update package versions in package-lock.json, whereas `npm ci` only installs the exact version listed in the package-lock.json. Changing package versions, even minor and patch updates, should never be something done at time of build but rather something that happens as a part of our maintenance and as an active choice.

* Yes, there are at least two dependencies that cannot be automatically built at this time with `tsc -b` (`lm-worker-internal` (or whatever we end up calling it), and `web-environment`). I think the best way forward for web-environment is to package the build steps for them in a `postinstall` script (though I hate the side-effect). `npm ci` at the root will then trigger the builds for it.

  But for `lm-worker-internal`, I am not sure that will do the job properly. Any change to its code would need a rebuild but that would not happen automatically with `tsc -b` in the web build, as it is manually appended to keymanweb.js (due to its wrapping). That's kinda annoying, so I am inclined to manually trigger its build from web's build.sh (it can still use `tsc -b` so the time cost should be low).

  We can change strategy later if it makes sense to do so!

File: web/source/build.sh
Comment (mcdurdin) [reply web/source/build.sh:57]: As the node_modules are all "hoisted" up with workspaces (though npm does not appear to use that lerna-term), the path change is needed. So it appears that the top-level `npm ci` did not finish, I think because the build agent had an outdated version of npm/node... (that's now resolved). Re-running the build, although I know it will still fail ;-)

**Review by jahorton — DISMISSED** [review]
Completed a first pass review through about everything.  Not a fan of all the needless whitespace changes given how many changes were already part of the PR, though I won't block merging on that front.  (I'm just complaining that they made the review process harder than it needed to be.)

The important section:  I've noted a few parts that seem odd and/or like development artifacts that may need cleaning up.  That said, do note that this review may be considered "in isolation" - I haven't even glanced at any of the other PRs in the chain yet, so maybe some (or all) of these points are addressed in later rungs of this PR ladder.

**Comment by jahorton** [comment]
> - merge web-environment and keyman-version

Proposed solution: #6625

> - cleanup "poor man's module mode" for tests from https://github.com/keymanapp/keyman/pull/6597.

Proposed solution: #6644

File: web/source/build.sh
Comment (jahorton) [reply web/source/build.sh:477]: Now addressed by #6644.

File: common/predictive-text/polyfills/object.values.js
Comment (mcdurdin) [reply common/predictive-text/polyfills/object.values.js:47]: This looks like an error. Will revert in upcoming PR.

File: common/predictive-text/polyfills/string.normalize.js
Comment (mcdurdin) [reply common/predictive-text/polyfills/string.normalize.js:124]: This looks like an error. Will revert in upcoming PR.

File: tsconfig.json
Comment (mcdurdin) [reply tsconfig.json:2]: Eventually, the idea is that we can run tsc -b from the base folder and it will build all the projects. Of course that is limited at present by the reality of the post-build steps we need. This follows a pattern from the typescript repo itself (and is [recommended](https://www.typescriptlang.org/docs/handbook/project-references.html#overall-structure) in the docs).

File: common/predictive-text/polyfills/object.values.js
Comment (mcdurdin) [reply common/predictive-text/polyfills/object.values.js:47]: My memory is hazy now, but I have a feeling that supportfor older versions of Android may have been what was motivating this change. This is certainly the wrong place to be doing it -- it should be in a separate PR to fixup support for older versions of Android. What I *think* happened is that I was working on these polyfills as a separate task and then they got caught up in the shuffle inadvertently.

* Arrow function support was [added in Chrome 45](https://caniuse.com/arrow-functions).
* `const` is similarly inconsistently supported in [versions earlier than Chrome 49](https://caniuse.com/const).

File: common/predictive-text/polyfills/string.normalize.js
Comment (mcdurdin) [reply common/predictive-text/polyfills/string.normalize.js:124]: See https://github.com/keymanapp/keyman/pull/6525#discussion_r883111326 (`let` only fully supported in [Chrome 49](https://caniuse.com/let))

File: web/source/kmwrotation.ts
Comment (mcdurdin) [reply web/source/kmwrotation.ts:109]: Because:

```
Argument of type 'Screen' is not assignable to parameter of type 'EventTarget'.
  Type 'Screen' is missing the following properties from type 'EventTarget': addEventListener, dispatchEvent, removeEventListenerts(2345)
```

I guess this is due to `mozorientationchange` being a non-standard event.

File: common/predictive-text/polyfills/object.values.js
Comment (mcdurdin) [reply common/predictive-text/polyfills/object.values.js:47]: OK. My memory was definitely hazy. I have investigated further. I reverted the changes locally and rebuilt. And here's what I got:

```
Minifying KeymanWeb...
WARNING - [SOURCEMAP_RESOLVE_FAILED] Failed to resolve sourcemap at ../intermediate/index.js.map: ..\intermediate\index.js.map

../intermediate/keymanweb.js:25883: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: const declaration.
const reduce = Function.bind.call(Function.call, Array.prototype.reduce);
^

../intermediate/keymanweb.js:25884: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: const declaration.
const isEnumerable = Function.bind.call(Function.call, Object.prototype.propertyIsEnumerable);
^

../intermediate/keymanweb.js:25885: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: const declaration.
const concat = Function.bind.call(Function.call, Array.prototype.concat);
^

../intermediate/keymanweb.js:25887: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: const declaration.
const keys = Object.keys;
^

../intermediate/keymanweb.js:25891: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: arrow function.
                return reduce(keys(O), (v, k) => concat(v, typeof k === 'string' && isEnumerable(O, k) ? [O[k]] : []), []);
                                       ^

../intermediate/keymanweb.js:25897: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: arrow function.
                return reduce(keys(O), (e, k) => concat(e, typeof k === 'string' && isEnumerable(O, k) ? [[k, O[k]]] : []), []);
                                       ^

../intermediate/keymanweb.js:26023: ERROR - [JSC_LANGUAGE_FEATURE] This language feature is only supported for ECMASCRIPT6 mode or better: let declaration.
      let chars = this.split('');
      ^
```

So... I am not sure why we weren't getting these errors earlier but am thinking perhaps we keep the changes?

File: common/predictive-text/worker/worker-interfaces.ts
Comment (mcdurdin) [reply common/predictive-text/worker/worker-interfaces.ts:35]: I've tried reverting this now and it looks fine. Was probably a temporary issue during the refactor.

File: common/predictive-text/worker/index.ts
Comment (mcdurdin) [reply common/predictive-text/worker/index.ts:386]: I was able to revert this change as well now.

File: common/predictive-text/worker/correction/distance-modeler.ts
Comment (mcdurdin) [reply common/predictive-text/worker/correction/distance-modeler.ts:11]: Yep, was able to revert this one as well.

**Comment by mcdurdin** [comment]
Review comments addressed in #6668

**Comment by mcdurdin** [comment]
# User Testing

This is a significant refactor. We need to do a regression test on this, sorry! Can we setup regression tests for Web, iOS and Android?

* TEST_REGRESSION_WEB: Run a full KeymanWeb regression test on this PR's artifacts in a new issue, report the overall result on this item.
* TEST_REGRESSION_IOS: Run a full Keyman for iPhone and iPad regression test on this PR's artifacts in a new issue, report the overall result on this item.
* TEST_REGRESSION_ANDROID: Run a full Keyman for Android regression test on this PR's artifacts in a new issue, report the overall result on this item.

Regression tests for Keyman Developer will be undertaken once the 🎢 Pull Request chain (#6629) is ready.

**Comment by mcdurdin** [comment]
> Changes have been addressed. Will convert to "approve" once user tests are positive.

Workflow-wise, we don't merge until all three of the following are satisfied:
* passing build
* user tests passed
* code reviews approved. 

So please go ahead and approve if you are satisfied with the code changes; there is no need to wait for user tests to pass. (Note: if changes are required as a result of user testing, then we can dismiss reviews as required.)

**Comment by mcdurdin** [comment]
# Test Results

* **TEST_REGRESSION_WEB (PASSED):** No issues detected. Regression test results in #6681.

**Review by jahorton — APPROVED** [review]
While I have seen a couple of interesting issues pop up from some of the app tests... absolutely none of it has been related to Web code, let alone changes from all the 🎡 work.

I'm comfortable giving this the LGTM at this point.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 16.0.12-alpha](https://keyman.com/downloads/releases/alpha/16.0.12)

**Pull request merged**
